text/0000-fields-in-traits.md

```diff
@@ -0,0 +1,800 @@
+will be significantly more expensive than a field assignment like
+`child.position = x`.
+
+This reveals one area where Rust
+[lacks expressiveness compared to C++ or Java][bs2]. A traditional
+class hierarchy allows you to insert fields (public or otherwise) at
+any point. These fields can be accessed with equal efficiency even if
+you don't know the exact runtime type of the object. In Rust, in
+contrast, you can only access fields if you have a concrete struct
+type; whenever you have a trait object, you can only use methods.
+
+[bs1]: http://smallcultfollowing.com/babysteps/blog/2015/05/05/where-rusts-enum-shines/
+[bs2]: http://smallcultfollowing.com/babysteps/blog/2015/05/29/classes-strike-back/
+
+#### The tradeoff: flexibility vs performance and safety
+
+There is a tradeoff here. Rust traits, like Java interfaces, were
+originally designed for maximum abstraction in mind. On that axis,
+accessors are clearly superior, since they allow a "field access" like
+`position()` or `set_position()` to require arbitrary computation or
+have arbitrary side effects. However, it is precisely that flexibility
+which leads to the two downsides cited above:
+
+1. The borrow checker must be more conservative because it does not,
+   and cannot, know what that two accessors like `children()` and
+   `set_position()` affect disjoint sets of fields.<sup>[1](#endnote1)</sup>
+2. The code generated must use a virtual call because it does not,
+   and cannot, know how any particular object plans to implement
+   `set_children()`.
+
+In a truly generic trait intended for implementation by arbitrary
+other types, this tradeoff for maximum flexibility is probably still a
+good idea. However, there are many cases where the flexibility to make
+accessors do arbitrary things (as opposed to simply access fields) is
+really not needed. Examples include traits that are local to a project
+(as opposed to public interfaces) as well as
+[attempts to model class-like hierarchies][aturon]. It seems best to
+leave this decision in the hands of the trait designer.
+
+[aturon]: http://aturon.github.io/blog/2015/09/18/reuse/
+
+# Detailed design
+[design]: #detailed-design
+
+### Fields in traits
+
+We extend traits with an optional "field block", which has the same
+structure as the contents of a struct (that is, a comma-separated list
+of fields; this RFC also adds the notation of a embedded prefix,
+described below). If following by further items, the Field block is
+terminated by a semicolon, but otherwise the semicolon may be omitted
+if the field block is the last item in the trait.
+
+```rust
+trait Trait {
+    field1: Type1, // <-- fields within a block separated by commas.
+    field2: Type2  // <-- no semicolon needed if there are no more items
+}
+
+trait Trait {
+    field1: Type1,
+    field2: Type2; // <-- semicolon terminates the field block
+
+    fn foo();
+}
+```
+
+**NB.** It might be simpler to just require semicolons after fields, or else
```

Comment:
I think this would be preferable. Its more consistent with the syntax of traits and much less complicated to understand. Having to edit the struct fields to terminate with `;` is not terribly burdensome (especially since this substitution applied to the lines should work, if your editor supports that sort of thing: `s/,?$/;/`.